crates/bevy_ecs/src/schedule/executor/multi_threaded.rs

```diff
@@ -279,16 +270,63 @@ impl SystemExecutor for MultiThreadedExecutor {
             std::panic::resume_unwind(payload);
         }
 
-        debug_assert!(self.ready_systems.is_clear());
-        debug_assert!(self.running_systems.is_clear());
-        self.active_access.clear();
-        self.evaluated_sets.clear();
-        self.skipped_systems.clear();
-        self.completed_systems.clear();
+        debug_assert!(state.ready_systems.is_clear());
+        debug_assert!(state.running_systems.is_clear());
+        state.active_access.clear();
+        state.evaluated_sets.clear();
+        state.skipped_systems.clear();
+        state.completed_systems.clear();
     }
 
     fn set_apply_final_deferred(&mut self, value: bool) {
-        self.apply_final_deferred = value;
+        self.state.get_mut().unwrap().apply_final_deferred = value;
+    }
+}
+
+impl<'scope, 'env: 'scope, 'sys> Context<'scope, 'env, 'sys> {
+    fn system_completed(
+        &self,
+        system_index: usize,
+        res: Result<(), Box<dyn Any + Send>>,
+        system: &BoxedSystem,
+    ) {
+        // tell the executor that the system finished
+        self.environment
```

Comment:
This doesn't have to be done in this PR, but we have the SystemResult here, and we may be able to lock the executor, sending this through the completion queue when we can just pass it in via a function argument may be wasteful and may contribute to contention on the queue.

Perhaps we could only add the result to the queue if and only if we failed to get the lock on the executor state.

Reply:
Nevermind, just tried this myself, seems to deadlock.

Reply:
Yeah, there's a race condition if one thread fails to get the lock, then the other thread exits the executor, and then the first thread pushes to the queue.